table.go

```diff
@@ -57,6 +60,48 @@ func NewRoutingTable(bucketsize int, localID ID, latency time.Duration, m peerst
 	return rt
 }
 
+func (rt *RoutingTable) GetAllBuckets() []*Bucket {
+	rt.tabLock.RLock()
+	defer rt.tabLock.RUnlock()
+	return rt.Buckets
+}
+
+func (rt *RoutingTable) GenRandPeerID(bucketID int) (peer.ID, error) {
+	rt.tabLock.RLock()
+	bucketLen := len(rt.Buckets)
+	rt.tabLock.RUnlock()
+
+	var targetCpl int
+	if bucketID >= bucketLen-1 {
+		targetCpl = bucketLen
+	} else {
+		targetCpl = bucketID
+	}
+
+	// should give up after a fixed number of attempts so we don't take up too much time/cpu
+	for i := 0; i < 1000; i++ {
+		peerID, err := randPeerID()
+		if err != nil {
+			log.Debugf("failed to generate random peerID in bucket %d, error is %+v", bucketID, err)
+			continue
+		}
+		if CommonPrefixLen(ConvertPeerID(peerID), rt.local) == targetCpl {
+			return peerID, err
+		}
+	}
+	return "", ErrGenRandPeerIDFailed
+}
+
+func randPeerID() (peer.ID, error) {
+	r := rand.New(rand.NewSource(time.Now().UnixNano()))
+	buf := make([]byte, 16)
+	if _, err := io.ReadFull(r, buf); err != nil {
+		return "", err
+	}
+	h, _ := mh.Sum(buf, mh.SHA2_256, -1)
```

Comment:
From [above](https://github.com/libp2p/go-libp2p-kbucket/pull/38/files#r309922392):

> > In any event it still doesn't seem like we need the hash function inside of randPeerID, just the one in ConvertPeerID.
> 
> The issue is that we need to make a DHT request with this key. To make the DHT request, we need to know the non-hashed key.

Sorry for not making this clearer. We still have to hash to turn `peerID` into a kad appropriate hash. However, in `randPeerID` there's no reason we need to hash random bytes when we could just craft a multihash with some random bytes in it (e.g. `mh.Encode(buf, mh.ID)` might work well and buf only needs to be 8 bytes long since `2^8=256`)

Reply:
@ aschmahmann Good point !  We've done something similar now.

@Stebalien has come up with a **mapping m** such that 
```
Let crmh(k) = A crafted sha2-256 multihash where the hash digest is a uint64 k.
Then, m[first sixteen bits of sha2-256(crmh(k))] = k
```

I've rebased on top of his branch and changed the code accordingly. Do let me know what you think !

Reply:
Seems reasonable to me, clever that we can avoid sending some bits on the wire by just using a bunch of storage (proportional to global network size).